File: backend-spring/src/main/java/com/cronos/controller/CultivationController.java
package com.cronos.controller;

import com.cronos.dto.response.ApiResponse;
import com.cronos.entity.*;
import com.cronos.service.CultivationService;
import lombok.RequiredArgsConstructor;
import org.springframework.http.ResponseEntity;
import org.springframework.security.core.Authentication;
import org.springframework.web.bind.annotation.*;
import java.util.UUID;

@RestController
@RequestMapping("/api/cultivations")
@RequiredArgsConstructor
public class CultivationController {
    private final CultivationService cultivationService;

    @PostMapping
    public ResponseEntity<?> create(Authentication auth, @RequestBody CultivationCycle cycle) {
        UUID userId = (UUID) auth.getPrincipal();
        try {
            return ResponseEntity.ok(ApiResponse.success("Cultivation cycle created", cultivationService.createCycle(userId, cycle)));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @GetMapping
    public ResponseEntity<?> getMyCycles(Authentication auth) {
        return ResponseEntity.ok(ApiResponse.success("Cycles retrieved", cultivationService.getMyCycles((UUID) auth.getPrincipal())));
    }

    @GetMapping("/{id}")
    public ResponseEntity<?> getCycle(@PathVariable UUID id) {
        return ResponseEntity.ok(ApiResponse.success("Cycle retrieved", cultivationService.getCycle(id)));
    }

    @PutMapping("/{id}")
    public ResponseEntity<?> update(Authentication auth, @PathVariable UUID id, @RequestBody CultivationCycle req) {
        try {
            return ResponseEntity.ok(ApiResponse.success("Cycle updated", cultivationService.updateCycle((UUID) auth.getPrincipal(), id, req)));
        } catch (RuntimeException e) { return ResponseEntity.badRequest().body(ApiResponse.error(e.getMessage())); }
    }

    @PostMapping("/{cycleId}/feed-logs")
    public ResponseEntity<?> addFeedLog(Authentication auth, @PathVariable UUID cycleId, @RequestBody FeedLog log) {
        log.setCultivationCycleId(cycleId);
        return ResponseEntity.ok(ApiResponse.success("Feed log added", cultivationService.addFeedLog((UUID) auth.getPrincipal(), log)));
    }

    @GetMapping("/{cycleId}/feed-logs")
    public ResponseEntity<?> getFeedLogs(@PathVariable UUID cycleId) {
        return ResponseEntity.ok(ApiResponse.success("Feed logs retrieved", cultivationService.getFeedLogs(cycleId)));
    }

    @PostMapping("/{cycleId}/water-quality")
    public ResponseEntity<?> addWaterQuality(Authentication auth, @PathVariable UUID cycleId, @RequestBody WaterQualityLog log) {
        log.setCultivationCycleId(cycleId);
        return ResponseEntity.ok(ApiResponse.success("Water quality log added", cultivationService.addWaterQualityLog((UUID) auth.getPrincipal(), log)));
    }

    @GetMapping("/{cycleId}/water-quality")
    public ResponseEntity<?> getWaterQuality(@PathVariable UUID cycleId) {
        return ResponseEntity.ok(ApiResponse.success("Water quality logs retrieved", cultivationService.getWaterQualityLogs(cycleId)));
    }
}

File: backend/controllers/cultivation_controller.go
package controllers
import (
	"net/http"
	"cronos-backend/middleware"
	"cronos-backend/models"
	"cronos-backend/services"
	"cronos-backend/utils"
	"github.com/gin-gonic/gin"
	"github.com/google/uuid"
)
type CultivationController struct {
	CultivationService *services.CultivationService
}
func NewCultivationController(service *services.CultivationService) *CultivationController {
	return &CultivationController{CultivationService: service}
}
func (ctrl *CultivationController) CreateCycle(c *gin.Context) {
	userID := middleware.GetUserID(c)
	var cycle models.CultivationCycle
	if err := c.ShouldBindJSON(&cycle); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	if err := ctrl.CultivationService.CreateCycle(userID, &cycle); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusCreated, "Cultivation cycle created", cycle)
}
func (ctrl *CultivationController) GetMyCycles(c *gin.Context) {
	userID := middleware.GetUserID(c)
	cycles, err := ctrl.CultivationService.GetMyCycles(userID)
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Cultivation cycles retrieved", cycles)
}
func (ctrl *CultivationController) GetCycle(c *gin.Context) {
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid ID", nil)
		return
	}
	cycle, err := ctrl.CultivationService.GetCycle(id)
	if err != nil {
		utils.NotFound(c, "Cycle not found")
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Cycle retrieved", cycle)
}
func (ctrl *CultivationController) UpdateCycle(c *gin.Context) {
	userID := middleware.GetUserID(c)
	id, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid ID", nil)
		return
	}
	var req models.CultivationCycle
	if err := c.ShouldBindJSON(&req); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	cycle, err := ctrl.CultivationService.UpdateCycle(userID, id, &req)
	if err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Cycle updated", cycle)
}
func (ctrl *CultivationController) AddFeedLog(c *gin.Context) {
	userID := middleware.GetUserID(c)
	cycleID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid ID", nil)
		return
	}
	var log models.FeedLog
	if err := c.ShouldBindJSON(&log); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	log.CultivationCycleID = cycleID
	if err := ctrl.CultivationService.AddFeedLog(userID, &log); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusCreated, "Feed log added", log)
}
func (ctrl *CultivationController) GetFeedLogs(c *gin.Context) {
	cycleID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid ID", nil)
		return
	}
	logs, err := ctrl.CultivationService.GetFeedLogs(cycleID)
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Feed logs retrieved", logs)
}
func (ctrl *CultivationController) AddWaterQualityLog(c *gin.Context) {
	userID := middleware.GetUserID(c)
	cycleID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid ID", nil)
		return
	}
	var log models.WaterQualityLog
	if err := c.ShouldBindJSON(&log); err != nil {
		utils.BadRequest(c, "Invalid request body", err.Error())
		return
	}
	log.CultivationCycleID = cycleID
	if err := ctrl.CultivationService.AddWaterQualityLog(userID, &log); err != nil {
		utils.BadRequest(c, err.Error(), nil)
		return
	}
	utils.SuccessResponse(c, http.StatusCreated, "Water quality log added", log)
}
func (ctrl *CultivationController) GetWaterQualityLogs(c *gin.Context) {
	cycleID, err := uuid.Parse(c.Param("id"))
	if err != nil {
		utils.BadRequest(c, "Invalid ID", nil)
		return
	}
	logs, err := ctrl.CultivationService.GetWaterQualityLogs(cycleID)
	if err != nil {
		utils.InternalServerError(c, err.Error())
		return
	}
	utils.SuccessResponse(c, http.StatusOK, "Water quality logs retrieved", logs)
}
